### PR TITLE
Refactor debug-only GenTree flags.

### PR DESCRIPTION
**Pull request by pgavlin**
Move the four debug-only GenTree flags into their own field. This frees
up two bits across both the RyuJIT and the legacy backend.

### CODE REVIEW

**Comment by pgavlin** [comment]
@dotnet/jit-contrib @mikedn PTAL

This is part of #5999.

File: src/jit/gentree.h [hunk 1]
```diff
@@ -687,15 +681,7 @@ struct GenTree
     #define GTF_DONT_CSE        0x00004000  // don't bother CSE'ing this expr
     #define GTF_COLON_COND      0x00008000  // this node is conditionally executed (part of ? :)
 
-#if defined(DEBUG) && SMALL_TREE_NODES
-    #define GTF_NODE_LARGE      0x00010000
-    #define GTF_NODE_SMALL      0x00020000
-
-    // Property of the node itself, not the gtOper
-    #define GTF_NODE_MASK       (GTF_COLON_COND | GTF_MORPHED   | GTF_NODE_SMALL | GTF_NODE_LARGE )
```
Comment (mikedn) [inline src/jit/gentree.h]: Something bugs me about this. `GTF_NODE_MASK` is used to prevent copying these flags to other nodes but it contains only 3 of the 4 moved flags. I suspect that some places that deal with flag copying might need to be updated. For example, at the end of `gtClone` there is:
```
copy->gtFlags |= tree->gtFlags & ~GTF_NODE_MASK;
```
Shouldn't the following line be added
```
copy->gtDebugFlags |= tree->gtDebugFlags & ~(GTF_DEBUG_MORPHED | GTF_DEBUG_NODE_SMALL | GTF_DEBUG_NODE_LARGE);
```
to copy `GTF_DEBUG_VAR_CSE_REF`?

Though it's not clear why `GTFD_VAR_CSE_REF` would be copied in this case, maybe the existing code is wrong to begin with.

File: src/jit/gentree.h [hunk 2]
```diff
@@ -857,6 +836,14 @@ struct GenTree
 
     //----------------------------------------------------------------
 
+#if defined(DEBUG)
+    #define GTF_DEBUG_MORPHED         0x00000001  // the node has been morphed (in the global morphing phase)
```
Comment (mikedn) [inline src/jit/gentree.h]: Rename to `GTF_DEBUG_NODE_MORPHED` to signal the fact that it is node rather than tree specific and to make it consistent with `GTF_DEBUG_NODE_SMALL/LARGE`?

**Comment by mikedn** [comment]
I'm not sure about the `GTF_NODE_MASK` thing but `GTF_DEBUG_VAR_CSE_REF` seems to be used only in dumps so it's not much of an issue. LGTM

File: src/jit/gentree.h (hunk 1 above)
Comment (pgavlin) [inline src/jit/gentree.h]: Yes, I think you're correct. I'll run through the uses of `GTF_NODE_MASK`.

File: src/jit/gentree.h (hunk 2 above)
Comment (pgavlin) [inline src/jit/gentree.h]: Sure, that sounds good.

File: src/jit/gentree.h (hunk 1 above)
Comment (briansull) [inline src/jit/gentree.h]: I believe that we should get rid of the GTF_MORPHED flag as it isn't really very useful or even consistently set. It mostly hit us when you try to morph a node a second time in the global morph phase and that is something that we often need to do.

Comment (CarolEidt) [inline src/jit/gentree.h]: I agree with @briansull about the GTF_MORPHED flag - I supposed it prevents some small amount of unintended throughput loss from accidentally re-morphing nodes, but is a huge hassle if that's what you really want to do.
I also agree that it is orthogonal to this.

File: src/jit/flowgraph.cpp [hunk 3]
```diff
@@ -20590,15 +20590,15 @@ void                Compiler::fgDebugCheckFlags(GenTreePtr tree)
     {
         // Print the tree so we can see it in the log.
         printf("Missing flags on tree [%X]: ", tree);
-        GenTree::gtDispFlags(chkFlags & ~treeFlags);
+        GenTree::gtDispFlags(chkFlags & ~treeFlags, 0);
```
Comment (CarolEidt) [inline src/jit/flowgraph.cpp]: Just a suggestion - it might be nice to have a GTF_NONE and/or a GTF_DEBUG_NONE for cases like this, just to make it clearer from the caller's side what this is.

**Comment by CarolEidt** [comment]
LGTM

File: src/jit/flowgraph.cpp (hunk 3 above)
Comment (pgavlin) [inline src/jit/flowgraph.cpp]: Good suggestion. Done.

**Comment by briansull** [comment]
LGTM

File: src/jit/gentree.h (hunk 1 above)
Comment (pgavlin) [inline src/jit/gentree.h]: Filed https://github.com/dotnet/coreclr/issues/6013 for follow-up.
